src/iter.rs

```diff
@@ -3,8 +3,13 @@ use smallvec::SmallVec;
 use crate::{Nucleotide, NucleotideAmbiguous, NucleotideLike};
 
 /// Helper trait to support iters regardless of whether their items are by-ref or by-value
-pub trait ToNucleotideLike {
-    type NucleotideType: NucleotideLike;
+pub trait ToNucleotideLike
+where
+    // It's much easier for the compiler to reason about chained Nucleotides iterator
+    // adapters if we explicitly state that ToNucleotideLike is idempotent.
+    Self::NucleotideType: ToNucleotideLike<NucleotideType = Self::NucleotideType>,
```

Comment:
Q since I'm not familiar with this trick: Does this just give better error messages? Faster compiles? Better codegen (presumably not)?

Reply:
It's a bug fix that I discovered I needed when implementing the `ForwardOrRcCodons` enum. When trying to impl `Iterator` for `ForwardOrRcCodons`, I ran into the problem that the compiler wanted me to add a large ugly `where` clause to prove that both `Codons<I>` and `Codons<Complement<Rev<I>>` iterated over the same type of item, whereas that feels like it should be obviously true given that RC shouldn't really change what something is iterating over. It took a bit of debugging to figure out why the compiler wasn't automatically inferring that: Technically you could add a bunch of `ToNucleotideLike` definitions in a big loop that never really converges to a particular value, so the compiler was correct that `Codons<I>` and `Codons<Complement<Rev<I>>` weren't guaranteed to iterate over the same type of item. Adding this fixes that and some other situations I discovered during debugging where chained `NucleotideIter` adapters weren't being recognized as nucleotide iterators.